### ISSUE TITLE
Unsound Uninitialized Memory

### ISSUE THREAD

**Issue by RustyYato** [SamuelFischerCode/vec-array#1]
https://github.com/SamuelFischerCode/vec-array/blob/ec034c8836f2f1de7fcbcff9c4669f39a967f905/src/vec.rs#L23

You can't set a generic `T` to uninitialzed even if you don't read those values, because of types lile bool or references which have illegal values (bool must be 0 or 1, references must not be null)
To correctly handle this, this, you should use `[MaybeUninit<T>; CAP]` instead.

**Comment by RustyYato** [comment]
Hey, I saw your post on Reddit and since you said you were learning, I thought I would take a peek. The code looks good, just this one issue stopping it from great!


https://www.reddit.com/r/rust/comments/12zjfpj/i_made_a_simple_crate_which_is_like_a_vector_but/

**Comment by shinmao** [comment]
@RustyYato Hi I saw this issue and also tried to figure out how to fix it.
> You can't set a generic T to uninitialized

Actually, the unsound implementation lies in [this line](https://github.com/SamuelFischerCode/vec-array/blob/ec034c8836f2f1de7fcbcff9c4669f39a967f905/src/vec.rs#LL52C54-L52C54), right?

However, I still don't understand how `[MaybeUninit<T>; CAP]` could help us fix invalid type on bool or references. Would you mind explain more about this?

**Comment by RustyYato** [comment]
> Actually, the unsound implementation lies in [this line](https://github.com/SamuelFischerCode/vec-array/blob/ec034c8836f2f1de7fcbcff9c4669f39a967f905/src/vec.rs#LL52C54-L52C54), right?

Yes, that's the line that will cause UB. Try running it under [miri](https://github.com/rust-lang/miri). With something like `VecArray<bool, 1>`.

> However, I still don't understand how `[MaybeUninit<T>; CAP]` could help us fix invalid type on bool or references. Would you mind explain more about this? 

Sure! Let's take a specific example to make the explanation easier. `VecArray<bool, 1>` which currently contains a `[bool; 1]`. When you set the array to uninitialized you are creating an invalid state for that `bool` (because `bool` must be either `0` or `1`, and uninitialized memory is neither of those). This is very bad because it gives LLVM (the compiler backend that Rust currently uses) free license to miscompile your code. The reason is, LLVM was guaranteed that bools must be `0` or `1`. However, you have produced (through `unsafe`) a bool that doesn't fit the requirements.

So how does `MaybeUninit<T>` make this better? The memory representation of `MaybeUninit<T>` has no requirements. It can be initialized to any value or left uninitialized altogether.  For example, `MaybeUninit<bool>` can have the value `0`, `1`, `2`, `3`, or uninit, etc. For now, treat `MaybeUninit` as special in this regard, the rules around what other types are allowed to be uninitialized are in flux[^1]. So altogether, by using `[MaybeUninit<T>; CAPACITY]`  it's safe to write uninitialized memory to the array.

A few more notes, `MaybeUninit<T>` has the same layout as `T`, so it's safe to cast `*const|mut T` to `*const|mut MaybeUninit<T>`. However, this safety doesn't compose with other types. For example, it's not safe to cast `*const|mut Option<T>` to `*const|mut Option<MaybeUninit<T>>` because of [niches](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html?highlight=niche#niche). For example, `Option<bool>` is 1 byte, but `Option<MaybeUninit<bool>>` is 2 bytes.

For you, this means that all you would need to do is change the type of `arr` to `[MaybeUninit<T>; CAPACITY]`  and the functions `VecArray::as_ptr` and `VecArray::as_mut_ptr` to be correct.

[^1]:  See https://github.com/rust-lang/unsafe-code-guidelines for work on specifying the Rust memory model.

**Comment by RustyYato** [comment]
I highly recommend reading the docs for `MaybeUninit`, it goes over this in far more detail and with a lot more examples

https://doc.rust-lang.org/std/mem/union.MaybeUninit.html

**Comment by shinmao** [comment]
Hi @RustyYato, thanks for the response with so much details.

```rust
pub struct VecArray<T, const CAP: usize> {
       arr: [T; CAP],
       ...
}
.....
let mut va: VecArray<T, 1> = VecArray::new();
```
Based on the definition of [invalid type in rustonomicon](https://doc.rust-lang.org/nomicon/what-unsafe-does.html), are there any types that can be valid with uninitialized state (in addition to `MaybeUninit`)? I think not only bool, but also signed/unsigned integer or other types can't work here, is it correct?

**Comment by RustyYato** [comment]
Iirc, integers are unspecified right now, so best treat them like it is UB to have uninit integers. But that may be relaxed in the future.

Other high profile cases are references, which can't be null and enum discriminants which must be in one of the enum variants.

**Comment by shinmao** [comment]
Yes, I just found the code in the docs of `MaybeUninit`:
```rust
let x: i32 = unsafe { MaybeUninit::uninit().assume_init() }; // undefined behavior! ⚠️
```
However, what makes me confused is the description in [rustonomicon](https://doc.rust-lang.org/nomicon/checked-uninit.html):
> Like C, all stack variables in Rust are uninitialized until a value is explicitly assigned to them.

Then, it means that before we explicitly assign value to integers, they are in invalid state, isn't it?

==========================================================================

Update

Sorry, I think I figure out the question above.
The delayed initialization itself works based on `MaybeUninit`.

**Comment by shinmao** [comment]
Hi @RustyYato, I found that some library authors claimed that it is safe if they immediately write value to the uninitialized memory just like following:

```rust
// safe function
fn() {
  let mut x: [T, N] = unsafe { MaybeUninit::uninit().assume_init() };
  // Write value into T immediately...
}
```

From my understanding, it should still be UB at the line of `assume_init()` no matter what happen following. How do you think about it?

**Comment by RustyYato** [comment]
Yep that's still UB
